File: internal/usecases/dto/question.go
package dto

import "strconv"

// =====================
// QuestionForm は質問のフォーム DTO である。DeletedAt は論理削除日時を表す。
type QuestionForm struct {
	ID               int64                 `json:"id"`
	OriginQuestionID *string               `json:"originQuestionId,omitempty"`
	SupportID        *int64                `json:"supportId,omitempty"`
	Title            string                `json:"title"`
	Content          string                `json:"content"`
	Due              *string               `json:"due,omitempty"`
	CreatedAt        *string               `json:"createdAt,omitempty"`
	UpdatedAt        *string               `json:"updatedAt,omitempty"`
	DeletedAt        *string               `json:"deletedAt,omitempty"`
	OriginQuestion   *QuestionForm         `json:"originQuestion,omitempty"`
	SubQuestions     []QuestionForm        `json:"subQuestions,omitempty"`
	Support          *SupportForm          `json:"support,omitempty"`
	Answers          []AnswerForm          `json:"answers,omitempty"`
	Memos            []MemoForm            `json:"memos,omitempty"`
	Tags             []TagForm             `json:"tags,omitempty"`
	EscalationsFrom  []EscalationForm      `json:"escalationsFrom,omitempty"`
	EscalationsTo    []EscalationForm      `json:"escalationsTo,omitempty"`
	RelatedQuestions []RelatedQuestionForm `json:"relatedQuestions,omitempty"`
	SenderTalks      []SenderTalkForm      `json:"senderTalks,omitempty"`
}

// OriginQuestionIDInt64 は OriginQuestionID 文字列を int64 に変換する。nil または空文字の場合は -1 を返す。
//
// return:
//   - int64: 元質問 ID（未設定・変換失敗時は -1）
func (f QuestionForm) OriginQuestionIDInt64() int64 {
	if f.OriginQuestionID == nil || *f.OriginQuestionID == "" {
		return -1
	}
	if val, err := strconv.ParseInt(*f.OriginQuestionID, 10, 64); err == nil {
		return val
	}
	return -1
}

File: static/js/models/Question.js
/**
 * Question model class.
 */

import { Support } from '/static/js/models/Support.js';
import { Answer } from '/static/js/models/Answer.js';
import { Memo } from '/static/js/models/Memo.js';
import { Tag } from '/static/js/models/Tag.js';
import { Notice } from '/static/js/models/Notice.js';
import { Escalation } from '/static/js/models/Escalation.js';
import { RelatedQuestion } from '/static/js/models/RelatedQuestion.js';
import { SenderTalk } from '/static/js/models/SenderTalk.js';
import { formToApiJson } from '/static/js/models/helpers.js';

export class Question {
  /**
   * @param {number}         id
   * @param {number|null}    originQuestionId
   * @param {number|null}    supportId
   * @param {string}         title
   * @param {string}         content
   * @param {string|null}    due              ISO8601 文字列 or null
   * @param {string}         createdAt        ISO8601 文字列
   * @param {string|null}    updatedAt
   * @param {Question|null}  [originQuestion]
   * @param {Question[]}     [subQuestions]
   * @param {Support|null}   [support]
   * @param {Answer[]}       [answers]
   * @param {Memo[]}         [memos]
   * @param {Tag[]}          [tags]
   * @param {Notice[]}       [notices]
   * @param {Escalation[]}   [escalationsFrom]
   * @param {Escalation[]}   [escalationsTo]
   * @param {RelatedQuestion[]} [relatedQuestions]
   * @param {SenderTalk[]} [senderTalks]
   */
  constructor(
    id,
    originQuestionId,
    supportId,
    title,
    content,
    due,
    createdAt,
    updatedAt,
    originQuestion = null,
    subQuestions = [],
    support = null,
    answers = [],
    memos = [],
    tags = [],
    notices = [],
    escalationsFrom = [],
    escalationsTo = [],
    relatedQuestions = [],
    senderTalks = [],
  ) {
    this.id = id;
    this.originQuestionId = originQuestionId;
    this.supportId = supportId;
    this.title = title;
    this.content = content;
    this.due = due ? new Date(due) : null;
    this.createdAt = new Date(createdAt);
    this.updatedAt = updatedAt ? new Date(updatedAt) : null;
    this.originQuestion = originQuestion;
    this.subQuestions = subQuestions;
    this.support = support;
    this.answers = answers;
    this.memos = memos;
    this.tags = tags;
    this.notices = notices;
    this.escalationsFrom = escalationsFrom;
    this.escalationsTo = escalationsTo;
    this.relatedQuestions = relatedQuestions;
    this.senderTalks = senderTalks;
  }

  /** @param {Object} json @returns {Question} */
  static fromJSON(json) {
    let oid = json.originQuestionId ?? null;
    if (oid !== null && typeof oid === 'string' && oid !== '') {
      oid = parseInt(oid, 10);
    }
    return new Question(
      json.id,
      oid,
      json.supportId ?? null,
      json.title,
      json.content,
      json.due ?? null,
      json.createdAt,
      json.updatedAt ?? null,
      json.originQuestion ? Question.fromJSON(json.originQuestion) : null,
      (json.subQuestions ?? []).map(Question.fromJSON),
      json.support ? Support.fromJSON(json.support) : null,
      (json.answers ?? []).map(Answer.fromJSON),
      (json.memos ?? []).map(Memo.fromJSON),
      (json.tags ?? []).map(Tag.fromJSON),
      (json.notices ?? []).map(Notice.fromJSON),
      (json.escalationsFrom ?? []).map(Escalation.fromJSON),
      (json.escalationsTo ?? []).map(Escalation.fromJSON),
      (json.relatedQuestions ?? []).map(RelatedQuestion.fromJSON),
      (json.senderTalks ?? []).map(SenderTalk.fromJSON),
    );
  }

  /** @returns {Object} */
  static toModel(form) {
    return formToApiJson(form);
  }
}
